File: 2022/day_02/puzzle.js
const { data_test } = require('./data_test.js');
const { puzzle_data } = require('./puzzle_data.js');

const ROCK = 1
const PAPER = 2
const SCISSORS = 3
const LOSE = 0
const DRAW = 3
const WIN = 6

function compute_data(data) {
    var res = []

    var split = data.split("\n");
    for (var i=0; i<split.length; i++) {
        res.push(split[i].split(" "))
    }
    return res;
}

function part1(data) {
    let res = 0
    for (let i=0; i<data.length; i++) {
        let game = data[i]
        let matchRes = 0
        if (game[0] === "A") {
            if (game[1] === "X") {
                matchRes = DRAW + ROCK
            } else if (game[1] === "Y") {
                matchRes = WIN + PAPER
            } else if (game[1] === "Z") {
                matchRes = LOSE + SCISSORS
            }
        } else if (game[0] === "B") {
            if (game[1] === "X") {
                matchRes = LOSE + ROCK
            } else if (game[1] === "Y") {
                matchRes = DRAW + PAPER
            } else if (game[1] === "Z") {
                matchRes = WIN + SCISSORS
            }
        } else if (game[0] === "C") {
            if (game[1] === "X") {
                matchRes = WIN + ROCK
            } else if (game[1] === "Y") {
                matchRes = LOSE + PAPER
            } else if (game[1] === "Z") {
                matchRes = DRAW + SCISSORS
            }
        }
        res += matchRes
    }
    return res
}

function part2(data) {
    let res = 0
    for (let i=0; i<data.length; i++) {
        let game = data[i]
        let matchRes = 0
        if (game[0] === "A") {
            if (game[1] === "X") {
                matchRes = LOSE + SCISSORS
            } else if (game[1] === "Y") {
                matchRes = DRAW + ROCK
            } else if (game[1] === "Z") {
                matchRes = WIN + PAPER
            }
        } else if (game[0] === "B") {
            if (game[1] === "X") {
                matchRes = LOSE + ROCK
            } else if (game[1] === "Y") {
                matchRes = DRAW + PAPER
            } else if (game[1] === "Z") {
                matchRes = WIN + SCISSORS
            }
        } else if (game[0] === "C") {
            if (game[1] === "X") {
                matchRes = LOSE + PAPER
            } else if (game[1] === "Y") {
                matchRes = DRAW + SCISSORS
            } else if (game[1] === "Z") {
                matchRes = WIN + ROCK
            }
        }
        res += matchRes
    }
    return res
}


let data = compute_data(puzzle_data)
console.log('Result for puzzle 1 = ', part1(data))
console.log('Result for puzzle 2 = ', part2(data))

File: 2024/day_11/puzzle.py
from functools import cache

def read_data(filename):
    return open(f"{filename}").read().strip()

###############################
## Construct data for solver ##
###############################
def constructData(data):
    m = []
    for num in data.split(" "):
        m.append(int(num))
    return m


###########
## UTILS ##
###########
def blink(num):
    res = []
    s = f'{num}'
    l = len(s)
    if num == 0:
        res.append(1)
    elif l%2 == 0:
        res.append(int(s[0:l//2]))
        res.append(int(s[l//2:]))
    else:
        res.append(num * 2024)
    return res

@cache
def count(num, nb):
    if nb == 0:
        return 1
    return sum(count(n, nb - 1) for n in blink(num))


###################
## Solver part 1 ##
###################
def part1(data):
    res = sum(count(num, 25) for num in data)
    return res


###################
## Solver part 2 ##
###################
def part2(data):
    res = sum(count(num, 75) for num in data)
    return res


##########
## MAIN ##
##########
# data = read_data("data_test.txt")
data = read_data("puzzle_data.txt")
print('Result for puzzle 1 = ', part1(constructData(data)))
print('Result for puzzle 2 = ', part2(constructData(data)))